python/tank_vendor/shotgun_authentication/shotgun_authenticator.py

```diff
@@ -0,0 +1,239 @@
+
+
+class ShotgunAuthenticator(object):
+    """
+    Shotgun Authentication
+    ----------------------
+
+    This class is used to help maintain an authenticated Shotgun User session
+    across multiple application launches and environments. By default, the
+    library is not tied to any particular Shotgun site - you can use it to
+    produce an authenticated user for any site of their choosing.
+
+    The library is essentially a series of factory methods, all returning
+    ShotgunUser derived instances. This instance represents an established user
+    in Shotgun. You can serialize this object and pass it around, etc. The
+    create_sg_connection() method returns a shotgun instance based on the
+    credentials of this user. It wraps around a Shotgun connection and traps
+    authentication errors so that whenever the Shotgun connection has expired,
+    it is automatically renewed by prompting the user to type in their password.
+    Whenever QT is available, a dialog is shown to aid in this prompting.
+
+    If you want to customize any of the logic of how the authentication
+    stores values, handles defaults or manages the behaviour in general,
+    implement an DefaultsManager class and pass it to the constructor of the
+    ShotgunAuthenticator object.
+    """
+
+    def __init__(self, defaults_manager=None):
+        """
+        Constructor
+
+        :param defaults_manager: A DefaultsManager object that defines the basic
+                                 behaviour of this authenticator. If omitted,
+                                 the default, built-in DefaultsManager will be
+                                 used.
+        """
+        self._defaults_manager = defaults_manager or DefaultsManager()
+
+    def get_saved_user(self):
+        """
+        Returns the currently saved user for the default site.
+
+        :returns: A ShotgunUser derived object or None if no saved user has been found.
+        """
+        host = self._defaults_manager.get_host()
+        # No default host, no so saved user can be found.
+        if not host:
+            return None
+        return user.SessionUser.get_saved_user(
+            host,
+            self._defaults_manager.get_http_proxy()
+        )
+
+    def clear_saved_user(self):
+        """
+        Removes the saved user's credentials from disk for the default host. The
+        next time the ShotgunAuthenticator.get_saved_user method is called,
+        None will be returned.
+
+        :returns: If a user was cleared, the user object is returned, None otherwise.
+        """
+        host = self._defaults_manager.get_host()
+        # No default host, no so saved user can be found.
+        if not host:
+            return None
+        sg_user = user.SessionUser.get_saved_user(
+            host,
+            self._defaults_manager.get_http_proxy()
+        )
+        if sg_user:
+            user.SessionUser.clear_saved_user(host)
```

Comment:
do we need all the checks? Could it not just call `user.SessionUser.clear_saved_user(host)` ?